viewing_party/party.py

```diff
@@ -1,23 +1,139 @@
+    except(UnboundLocalError):
+        pass
+    return user_data
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    total = 0
+    num_iterations = 0
+    for i in user_data["watched"]:
+        num_iterations += 1
+        total += i['rating']
+    try:
+        average = total / num_iterations
+    except(ZeroDivisionError):
+        return 0.0
+    return average
+    
+def get_most_watched_genre(user_data):
+    if user_data["watched"] == []:
+        return None
+    genre_tally = {}
+    for movie in user_data["watched"]:
+        try:
+            genre_tally[movie["genre"]] += 1
+        except (KeyError):
+            genre_tally[movie["genre"]] = 1
+    most_watched = (None, 0)
+    for item in genre_tally.items():
+        if item[1] > most_watched[1]:
+            most_watched = (item[0], item[1])
+    return most_watched[0]
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    friend_movies = set()
+    for friend in user_data["friends"]:
+        for friend_movie in friend["watched"]:
+            friend_movies.add(friend_movie["title"])
+    unique_watched = []
+    for movie in user_data["watched"]:
+        if movie["title"] not in friend_movies:
+            unique_watched.append(movie)
+    return unique_watched
         
+def get_friends_unique_watched(user_data):
+    user_movie_titles = set()
+    for user_movie in user_data["watched"]:
+        user_movie_titles.add(user_movie["title"])
+    friend_unique_titles = set()
+    friends_unique = []
+    for friend in user_data["friends"]:
+        for friend_movie in friend["watched"]:
+            if friend_movie["title"] in user_movie_titles:
+                continue
+            elif friend_movie["title"] in friend_unique_titles:
+                continue
+            else:
+                friend_unique_titles.add(friend_movie["title"])
+                friends_unique.append(friend_movie)
+    return friends_unique
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    recommended_movies = []
+    friends_unique = get_friends_unique_watched(user_data)
+    for movie in friends_unique:
+        if movie["host"] in user_data["subscriptions"]:
+            recommended_movies.append(movie)
+    return recommended_movies
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    if len(user_data["watched"]) == 0:
+        return []
+    max_genre = get_most_watched_genre(user_data)
+    recs_by_genre = []
+    for rec in get_available_recs(user_data):

```

Comment:
`get_available_recs` checks if the user has each movie's host service. This makes our recommendations a little more restrictive than the README requires. You don't have to write out code, but consider: how we would change things if we wanted to recommend movies that are in our favorites but we may not have the host service for?